data/transactions/transactions_gen.go

```diff
@@ -0,0 +1,2652 @@
+	o = append(o, 0x80|uint8(zb0001Len))
+	if zb0001Len == 0 {
+		return
+	}
+	if (zb0001Mask & 0x2) == 0 { // if not empty
+		// string "amt"
+		o = append(o, 0xa3, 0x61, 0x6d, 0x74)
+		o, err = z.Amount.MarshalMsg(o)
+		if err != nil {
+			err = msgp.WrapError(err, "Amount")
+			return
+		}
+	}
+	if (zb0001Mask & 0x4) == 0 { // if not empty
+		// string "close"
+		o = append(o, 0xa5, 0x63, 0x6c, 0x6f, 0x73, 0x65)
+		o, err = z.CloseRemainderTo.MarshalMsg(o)
+		if err != nil {
+			err = msgp.WrapError(err, "CloseRemainderTo")
+			return
+		}
+	}
+	if (zb0001Mask & 0x8) == 0 { // if not empty
+		// string "rcv"
+		o = append(o, 0xa3, 0x72, 0x63, 0x76)
+		o, err = z.Receiver.MarshalMsg(o)
+		if err != nil {
+			err = msgp.WrapError(err, "Receiver")
+			return
+		}
+	}
+	return
+}
+
+// UnmarshalMsg implements msgp.Unmarshaler
+func (z *PaymentTxnFields) UnmarshalMsg(bts []byte) (o []byte, err error) {
+	var field []byte
+	_ = field
+	var zb0001 uint32
+	zb0001, bts, err = msgp.ReadMapHeaderBytes(bts)
+	if err != nil {
+		err = msgp.WrapError(err)
+		return
+	}
+	for zb0001 > 0 {
+		zb0001--
+		field, bts, err = msgp.ReadMapKeyZC(bts)
+		if err != nil {
+			err = msgp.WrapError(err)
+			return
+		}
+		switch msgp.UnsafeString(field) {
+		case "rcv":
+			bts, err = z.Receiver.UnmarshalMsg(bts)
+			if err != nil {
+				err = msgp.WrapError(err, "Receiver")
+				return
+			}
+		case "amt":
+			bts, err = z.Amount.UnmarshalMsg(bts)
+			if err != nil {
+				err = msgp.WrapError(err, "Amount")
+				return
+			}
+		case "close":
+			bts, err = z.CloseRemainderTo.UnmarshalMsg(bts)
+			if err != nil {
+				err = msgp.WrapError(err, "CloseRemainderTo")
+				return
+			}
+		default:
+			err = msgp.ErrNoField(msgp.UnsafeString(field))
+			if err != nil {
+				err = msgp.WrapError(err)
+				return
+			}
+		}
+	}
+	o = bts
+	return
+}
+
+// Msgsize returns an upper bound estimate of the number of bytes occupied by the serialized message
+func (z *PaymentTxnFields) Msgsize() (s int) {
+	s = 1 + 4 + z.Receiver.Msgsize() + 4 + z.Amount.Msgsize() + 6 + z.CloseRemainderTo.Msgsize()
+	return
+}
+
+// MsgIsZero returns whether this is a zero value
+func (z *PaymentTxnFields) MsgIsZero() bool {
+	return (z.Receiver.MsgIsZero()) && (z.Amount.MsgIsZero()) && (z.CloseRemainderTo.MsgIsZero())
+}
+
+// MarshalMsg implements msgp.Marshaler
+func (z Payset) MarshalMsg(b []byte) (o []byte, err error) {
+	o = msgp.Require(b, z.Msgsize())
+	o = msgp.AppendArrayHeader(o, uint32(len(z)))
+	for za0001 := range z {
+		o, err = z[za0001].MarshalMsg(o)
+		if err != nil {
+			err = msgp.WrapError(err, za0001)
+			return
+		}
+	}
+	return
+}
+
+// UnmarshalMsg implements msgp.Unmarshaler
+func (z *Payset) UnmarshalMsg(bts []byte) (o []byte, err error) {
+	var zb0002 uint32
+	zb0002, bts, err = msgp.ReadArrayHeaderBytes(bts)
+	if err != nil {
+		err = msgp.WrapError(err)
+		return
+	}
+	if cap((*z)) >= int(zb0002) {
+		(*z) = (*z)[:zb0002]
+	} else {
+		(*z) = make(Payset, zb0002)
```

Comment:
For example, it looks like you can potentially force it to allocate space for `2**32 - 1` `SignedTxnInBlock`s here with a short message? `ReadArrayHeaderBytes` is [here](https://github.com/zeldovich/msgp/blob/master/msgp/read_bytes.go#L220).

Reply:
Also, I think this could panic on 32-bit systems when `zb0002` is > `2**31 - 1`. Check out https://play.golang.org/p/cvjxsE7N4s- (I think the go playground site is 32-bit, system int size is 4 bytes).

Reply:
What were you thinking of for bounding the allocation?  One possibility I'm considering is to ensure that all maps and arrays must be no larger than the remaining number of bytes available to decode.  So, if we cap the size of the message (elsewhere in the code), the decoder blowup will be somewhat bounded over that max message size.  Tighter bounds are probably possible, though trickier to get right.

Reply:
"no larger than the remaining number of bytes" sounds reasonable to me. Though I guess that could still give a blow up of approximately `(number of bytes being decoded) * (maximum struct size we decode slices of)`. Which could still be a lot... I'll check around the places we decode messages.

Reply:
In `rpcs/httpTxSync.go`, we `protocol.Decode` a slice of `transactions.SignedTxn`. On my machine, `unsafe.Sizeof(transactions.SignedTxn{})` is 1056. If I'm reading it correctly, the maximum length body that gets decoded is 1,000,016 bytes. So you could allocate about 1GB of space per sync response there.

Reply:
In catchup, we use the `WsFetcherService` to fetch blocks + certs, which has a 4MB response limit. `unsafe.Sizeof(transactions.SignedTxnInBlock{})` is 1096. So during catchup you can cause a ~4GB allocation. Since we run catchup goroutines in parallel, that could add up pretty quickly if the responses are timed together.